include/exec/static_thread_pool.hpp

```diff
@@ -424,6 +422,263 @@ namespace exec {
         return bulk_sender_t<S, Shape, Fn>{*sch.pool_, (S&&) sndr, shape, (Fn&&) fun};
       }
 
+      template <class SenderId, class ReceiverId, class InitT, class RedOp>
+      struct reduce_shared_state {
+        using Sender = stdexec::__t<SenderId>;
+        using Receiver = stdexec::__t<ReceiverId>;
+
+        struct reduction_task : task_base {
+          reduce_shared_state* sh_state_;
+
+          reduction_task(reduce_shared_state* sh_state)
+            : sh_state_(sh_state) {
+            task_base::__execute = [](task_base* t, const std::uint32_t tid) noexcept {
+              auto& sh_state = *static_cast<reduction_task*>(t)->sh_state_;
+              auto total_threads = sh_state.num_threads_available();
+              auto& partial = sh_state.partials_[tid];
+
+              auto reducer = [&](auto& input) {
+                const auto [first, last] = even_share(std::ranges::size(input), tid, total_threads);
+                if (first != last) {
+                  auto begin = std::ranges::begin(input) + first;
+                  partial = std::reduce(
+                    std::next(begin), std::ranges::begin(input) + last, *begin, sh_state.redop_);
+                }
+              };
+
+              auto finalize = [&](auto&) {
+                auto result = std::reduce(
+                  std::ranges::begin(sh_state.partials_),
+                  std::ranges::end(sh_state.partials_),
+                  sh_state.init_,
+                  sh_state.redop_);
+                // deallocate
+                sh_state.partials_ = {};
+                stdexec::set_value((Receiver&&) sh_state.receiver_, std::move(result));
+              };
+
+              sh_state.apply(reducer);
+
+              const bool is_last_thread = sh_state.finished_threads_.fetch_add(1)
+                                       == (total_threads - 1);
+              if (is_last_thread) {
+                sh_state.apply(finalize);
+              }
+            };
+          }
```

Comment:
We don't appear to be handling exceptions that may be thrown from the reduction function.

Reply:
I know. I left exception handling out, just wanted stuff to work first. I'll add it.